tensorflow/core/kernels/sparse_fill_empty_rows_op_gpu.cu.cc

```diff
@@ -0,0 +1,674 @@
+  return GpuLaunchKernel(func, config.block_count, config.thread_per_block, 0,
+                         device.stream(), config, to_pointers(args)...);
+}
+};  // namespace kernel_forward
+
+using kernel_forward::wrap_kernel_call;
+
+namespace functor {
+
+namespace {
+
+// Computes elements_per_row[0..dense_rows] and sets *rows_are_not_ordered to
+// true if the indices are not ordered by row.
+template <typename Tindex>
+__global__ __launch_bounds__(1024) void CountElementsPerRowKernel(
+    GpuLaunchConfig cfg, Tindex dense_rows, int rank, const Tindex* indices,
+    Tindex* elements_per_row, int* rows_are_not_ordered,
+    int* first_invalid_index) {
+  GPU_1D_KERNEL_LOOP(i, cfg.virtual_thread_count) {
+    Tindex row = indices[i * rank];
+    if (row < 0 || row >= dense_rows) {
+      GpuAtomicMin(first_invalid_index, i);
+      continue;
+    }
+    GpuAtomicAdd(&elements_per_row[row], 1);
+    if (i > 0 && row < indices[(i - 1) * rank]) {
+      // TODO(benbarsdell): Replace this with atomic_ref::store when available.
+      GpuAtomicMax(rows_are_not_ordered, 1);
+    }
+  }
+}
+
+template <typename Tindex>
+__global__ __launch_bounds__(1024) void CopyRowIndicesKernel(
+    GpuLaunchConfig cfg, int rank, const Tindex* indices, Tindex* row_indices) {
+  GPU_1D_KERNEL_LOOP(i, cfg.virtual_thread_count) {
+    row_indices[i] = indices[i * rank];
+  }
+}
+
+template <typename Tindex>
+__global__ __launch_bounds__(1024) void RangeInitKernel(GpuLaunchConfig cfg,
+                                                        Tindex start,
+                                                        Tindex delta,
+                                                        Tindex* out) {
+  GPU_1D_KERNEL_LOOP(i, cfg.virtual_thread_count) {
+    out[i] = start + i * delta;
+  }
+}
+
+// Sets empty_row_indicator[row] to whether the row is empty.
+template <typename Tindex>
+__global__ __launch_bounds__(1024) void ComputeEmptyRowIndicatorKernel(
+    GpuLaunchConfig cfg, const Tindex* elements_per_row,
+    bool* empty_row_indicator) {
+  GPU_1D_KERNEL_LOOP(row, cfg.virtual_thread_count) {
+    empty_row_indicator[row] = elements_per_row[row] == 0;
+  }
+}
+
+// Copies indices and values to output_indices and output_values, leaving space
+// in the output for the new elements that will be inserted wherever there is an
+// empty row.
+template <typename T, typename Tindex>
+__global__ __launch_bounds__(1024) void ScatterInputElementsKernel(
+    GpuLaunchConfig cfg, Tindex dense_rows, int rank,
+    const Tindex* input_index_map, const Tindex* indices, const T* values,
+    const Tindex* num_new_rows_before, Tindex* output_indices, T* output_values,
+    Tindex* reverse_index_map) {
+  GPU_1D_KERNEL_LOOP(i, cfg.virtual_thread_count) {
+    Tindex input_i = input_index_map ? input_index_map[i] : i;
+    Tindex row = indices[input_i * rank];
+    if (row < 0 || row >= dense_rows) {
+      continue;  // Ignore indices that are out of range
+    }
```

Comment:
Is this needed?  Won't we have returned an error if any indices were out of range?

Reply:
Good point, removed.